Add gcc9 config and remove ZeroInit macro

am11 opened the pull request:
<del>`MSLAYOUT` macro predates the recent gcc support, and was part of [initial port](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/debug/inc/common.h#L299) in CoreCLR.</del> (removing MSLAYOUT breaks i386 build)

experimenting with the idea that some of the cases in https://github.com/dotnet/coreclr/issues/24518 do not require `memset` in the current form.

am11 commented:
cc @jkotas
diff without whitespace change: https://github.com/dotnet/coreclr/pull/27605/files?w=1

noahfalk commented on src/inc/dacprivate.h:
```diff
@@ -97,14 +60,14 @@ struct MSLAYOUT DacpObjectData : ZeroInit<DacpObjectData>
     }
 };
 
-struct MSLAYOUT DacpExceptionObjectData : ZeroInit<DacpExceptionObjectData>
```
Is there some rationale I should apply about why removing the ZeroInit is safe? Superficially it appears that previously we were zero initting our data, now we are not zero initting, and maybe this has caused some bugs because the consuming code required the zero init for correctness?

My hunch is that it would be easier to define a default constructor for each type which does the zero init rather than prove via code review that the zero init wasn't required for correctness to begin with.

mikem8361 replied:
Even though this h file has "private" in its name it actually is the internal interface between SOS and the DAC.  SOS now resides in a different [repo](https://github.com/dotnet/diagnostics.git) so the alignment that MSLAYOUT does (or did) needs to be the same before/after this change.

I think the DAC API assumes that this structs are zero initialized when passed by SOS.

am11 replied:
I have initialized the fields of each struct:

* BOOL -> FALSE
* integers -> 0
* enums -> 0-index value
* arrays -> {}

during my testing, in one case, `static_assert` on top of this file was violated, which helped me validate the change in terms of: struct sizes are same before/after.

noahfalk commented:
cc @mikem8361 @sdmaclea

sdmaclea commented:
> MSLAYOUT macro predates the recent gcc support, and was part of initial port in CoreCLR.

This sentence is basically doesn't really tell me why you are removing the macro. "Predates" means it was supported before the recent gcc versions. Since the initial publish of CoreCLR open source was before recent gcc, the statement seems meaningless.

We definitely will want to control the layout of these structures for cross platform binary compatibility as mentioned in some of the comments you are removing.  How will that be controlled now?

Why not just change the definition of the MSLAYOUT macro rather than removing it?

mikem8361 commented:
I wouldn't include all the white space changes. It really makes it hard to see exactly what has changed.

noahfalk commented:
> for cross platform binary compatibility

We genuinely needed this requirement on Silverlight, but for .Net Core the comments have become outdated and misleading. For Silverlight we had a mac remote debugging scenario where mscordbi.dll ran on a Windows machine and these data types crossed a network interface. For .Net Core we expect mscordbi.dll to always run on the same machine as the debuggee using shared memory or a local pipe to connect them so these data types never cross the network. As long as mscordbi.dll and coreclr.dll from the same build agree on the layout I'd expect we are OK.

sdmaclea commented:
> static_assert on top of this file was violated, which helped me validate the change in terms of: struct sizes are same before/after.

So are you saying, that there is a static assert for each structure which has the MSLAYOUT attribute, so there is no risk to introduce unintended binary change for any target platform (arch + os)?

am11 commented:
Reverted whitespace (in favor of #27626) and MSLAYOUT changes.

@sdmaclea, sorry for the insufficient description. I started it as a draft PR, and ended up in this state.

Similar legacy and ms-specific macros were removed since the initial port, and I understood that there is general drive to make the code-base portable, ISO C/C++ standard conformant.

The (initial) idea behind removing `MSLAYOUT` is that in case of clang (which seems like it was not the supported compiler in pre-OSS era), this alignment is not required with current revision, so gcc (which is not fully supported post-OSS) might as well not need it.

What I failed to realize; clang was also using this attribute as LLVM also defines (gcc well documented) `__GNUC__` macro, in addition to its own `__llvm__`.

> no risk to introduce unintended binary change for any target platform (arch + os)?

Now that you have mentioned, I tried building my branch on Ubuntu i386 with clang, and found that the static assertion is being violated during the build. The fix was introduced to explicitly use `MSLAYOUT` attribute in dacprivate structs here: https://github.com/dotnet/coreclr/issues/8369. Therefore I have reverted the `MSLAYOUT` change. :)
